cobra/flux_analysis/deletion.py

```diff
@@ -0,0 +1,348 @@
+# -*- coding: utf-8 -*-
+
+import sys
+import multiprocessing
+import logging
+from warnings import warn
+from itertools import product
+from operator import attrgetter
+from collections import defaultdict
+from builtins import (map, dict)
+from future.utils import raise_
+
+from ..manipulation.delete import (find_gene_knockout_reactions)
+import cobra.util.solver as sutil
+
+try:
+    import scipy
+except ImportError:
+    moma = None
+else:
+    from . import moma
+
+try:
+    from pandas import DataFrame
+except:
+    DataFrame = None
+
+LOGGER = logging.getLogger(__name__)
+
+
+def infinite_defaultdict():
+    return defaultdict(infinite_defaultdict)
+
+
+def defaultdict_to_dict(def_dict):
+    if not isinstance(def_dict, defaultdict):
+        return def_dict
+    for k, v in def_dict.items():
+        def_dict[k] = defaultdict_to_dict(v)
+    return dict(def_dict)
+
+
+def _reactions_knockouts_with_restore(model, reactions):
```

Comment:
The 'restore' functionality is exactly provided by the model context manager. Any model modifications are reversed upon leaving the context. You can find examples [here](http://cobrapy.readthedocs.io/en/latest/getting_started.html?highlight=context#Making-changes-reversibly-using-models-as-contexts).

Reply:
that's true, though the speed issue was exactly here: it was much faster in my benchmarks to restore the simple case of few reactions fluxes like that than to use the context manager

Reply:
It should not really make a difference if you use nested contexts, like in the current implementation:

```Python
        with cobra_model as m:
            m.solver = solver
            for gene in gene_list:
                with m:
                    gene.knock_out()
                    obj_value = m.slim_optimize()
                    status_dict[gene.id] = m.solver.status
                    growth_rate_dict[gene.id] = obj_value
```